### PR TITLE
Freedom Ad Network Bid Adapter: initial release

### PR DESCRIPTION
**Pull request by freedomadnetworkdev**
## Type of change
- [x] New bidder adapter

## Description of change
Initial release of the Freedom Ad Network Bidder Adapter. The adapter requires setup before beginning. Please contact us at [info@freedomadnetwork.com](mailto:info@freedomadnetwork.com).

```
{
  bidder: 'freedomadnetwork',
  params: {
    placementId: 'e6203f1e-bd6d-4f42-9895-d1a19cdb83c8'
  }
}
```

## Other information
Documentation PR: https://github.com/prebid/prebid.github.io/pull/5560

### CODE REVIEW

File: modules/fanAdapter.js [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+  }
+
+  var screen = [
+    parseInt(window.screen.width),
+    parseInt(window.screen.height),
+    window.screen.availHeight,
+    window.screen.availWidth,
+  ];
+
+  var hash = [
```
Comment (patmmccann) [inline modules/fanAdapter.js:234]: You might find some more ideas for your fingerprint at https://amiunique.org/fingerprint or
https://github.com/duckduckgo/tracker-radar/blob/main/build-data/generated/api_fingerprint_weights.json :) We're not able to accept this pr

Comment (freedomadnetworkdev) [reply modules/fanAdapter.js]: I have removed the problematic code and applied all recommendations.

File: modules/fanAdapter.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  }
+}
+
+/**
+ * Generate stable user id
+ *
+ * @returns {String} userId
+ */
+function generateUserId() {
+  var hash = MD5(navigator.userAgent).toString();
```
Comment (patmmccann) [inline modules/fanAdapter.js:105]: why perform an expensive calculation on this instead of just sending it ?

Comment (freedomadnetworkdev) [reply modules/fanAdapter.js:105]: Server side expects an md5.

File: modules/fanAdapter.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
+import * as utils from '../src/utils.js';
+import MD5 from 'crypto-js/md5.js';
+import { ajax } from '../src/ajax.js';
+import { BANNER, NATIVE } from '../src/mediaTypes.js';
+import { config } from '../src/config.js';
+import { registerBidder } from '../src/adapters/bidderFactory.js';
+const BIDDER_CODE = 'freedomadnetwork';
```
Comment (patmmccann) [inline modules/fanAdapter.js:7]: you may want a short code alias

File: modules/fanAdapter.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    navigator.language || navigator.userLanguage;
+  return lang ? lang.split('-')[0] : 'en';
+}
+
+/**
+ * Get device info
+ *
+ * @returns {Object}
+ */
+function getDevice() {
```
Comment (patmmccann) [inline modules/fanAdapter.js:36]: all of this should come off the request object

File: modules/fanAdapter.js [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+  const payload = {
+    id: bid.bidId,
+    tmax: bidderRequest.timeout,
+    placements: [bid.params.placementId],
+    test: config.getConfig('debug') ? 1 : 0,
+    device: getDevice(),
+    at: 2,
+    user: {
+      coppa: config.getConfig('coppa') ? 1 : 0,
```
Comment (patmmccann) [inline modules/fanAdapter.js:75]: this is also on the request

File: modules/fanAdapter.js [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+ * @returns {Request}
+ */
+function buildBidRequest(bid, bidderRequest) {
+  const userId = getUserId(bid);
+
+  const payload = {
+    id: bid.bidId,
+    tmax: bidderRequest.timeout,
+    placements: [bid.params.placementId],
+    test: config.getConfig('debug') ? 1 : 0,
```
Comment (patmmccann) [inline modules/fanAdapter.js:71]: this is on the request

File: modules/fanAdapter.js [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    device: getDevice(),
+    at: 2,
```
Comment (patmmccann) [inline modules/fanAdapter.js:73]: this isnt true

File: modules/fanAdapter.js [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+  return {
+    method: 'POST',
+    url: BASE_URL + '/pb/req',
+    data: JSON.stringify(payload),
+    options: {
+      contentType: 'application/json',
+      withCredentials: false,
+      customHeaders: {
+        'Accept-Language': 'en;q=10',
+        'Authorization': 'Bearer ' + userId
```
Comment (patmmccann) [inline modules/fanAdapter.js:99]: this is odd

File: modules/fanAdapter.js (hunk 7 above)
Comment (freedomadnetworkdev) [reply modules/fanAdapter.js:73]: What do you mean? Should be a constant?

File: modules/fanAdapter.js (hunk 8 above)
Comment (freedomadnetworkdev) [reply modules/fanAdapter.js:99]: Unfortunately, I cannot change this. Server side requires this interface.

File: modules/fanAdapter.js (hunk 6 above)
Comment (freedomadnetworkdev) [reply modules/fanAdapter.js:71]: Based myself off of the https://github.com/prebid/Prebid.js/blob/master/modules/operaadsBidAdapter.js#L246 adapter. Is there a newer one to use as reference?

File: modules/fanAdapter.js (hunk 8 above)
Comment (patmmccann) [reply modules/fanAdapter.js:65]: why?

Comment (freedomadnetworkdev) [reply modules/fanAdapter.js:65]: I have no control over this part of the service.

**Review by patmmccann — APPROVED** [review]

**Pull request merged**